src/main/java/org/zalando/nakadi/service/subscription/zk/CuratorZkSubscriptionClient.java

```diff
@@ -317,4 +323,110 @@ public ZkSubscriptionNode getZkSubscriptionNodeLocked() {
 
         return subscriptionNode;
     }
+
+    @Override
+    public ZKSubscription subscribeForCursorsReset(final Runnable listener) {
+        final NodeCache cursorResetCache = new NodeCache(curatorFramework, resetCursorPath);
+        cursorResetCache.getListenable().addListener(() -> listener.run());
+
+        try {
+            cursorResetCache.start();
+        } catch (final Exception e) {
+            throw new NakadiRuntimeException(e);
+        }
+
+        return new ZKSubscription() {
+            @Override
+            public void refresh() {
+                try {
+                    cursorResetCache.rebuild();
+                } catch (final Exception e) {
+                    throw new NakadiRuntimeException(e);
+                }
+            }
+
+            @Override
+            public void cancel() {
+                try {
+                    cursorResetCache.getListenable().clear();
+                    cursorResetCache.close();
+                } catch (final IOException e) {
+                    throw new NakadiRuntimeException(e);
+                }
+            }
+        };
+    }
+
+    @Override
+    public boolean isCursorResetInProgress() {
+        try {
+            return curatorFramework.checkExists().forPath(resetCursorPath) != null;
+        } catch (final Exception e) {
+            // nothing in the path
+        }
+        return false;
+    }
+
+    @Override
+    public void resetCursors(final List<NakadiCursor> cursors, final Long timeout)
+            throws OperationTimeoutException, ZookeeperException {
+        ZKSubscription sessionsListener = null;
+        try {
+            // close subscription connections
+            curatorFramework.create().withMode(CreateMode.EPHEMERAL).forPath(resetCursorPath);
+
+            final String sessionsPath = MessageFormat.format("/nakadi/subscriptions/{0}/sessions", subscriptionId);
+            final Object lock = new Object();
+            sessionsListener = subscribeForSessionListChanges(() -> {
+                try {
+                    // wait until all connections are closed by nakadi
+                    if (curatorFramework.getChildren().forPath(sessionsPath).isEmpty()) {
```

Comment:
It is not allowed to make calls to curator from notification threads. 2-nd call will stuck.